### PR TITLE
Remote module execution

### PR DESCRIPTION
**Pull request by rjeffman**
All ansible-freeipa modules are required to run on an IPA server. The Ansible community modules use JSON API and allow the execution on other environments. This patch allows the execution of ansible-freeipa modules using IPA's Python API on an IPA client host

A new optional attribute, `ipa_context`, is added to the modules that allow the execution in a client context. The valid values for this attribute are server (default) and client. All current modules can be executed in a client context, but `ipavault` module, currently, which accepts the attribute but will fail if it is set to `server`.

Related to: https://bugzilla.redhat.com/show_bug.cgi?id=1918025

### CODE REVIEW

**Comment by rjeffman** [comment]
@varunmylaraiah if you can, please, test remote module execution against this branch.

**Comment by varunmylaraiah** [comment]
I have tested remote module execution MR against this branch, and all tests are pass.
https://gitlab.cee.redhat.com/identity-management/ansible-freeipa-tests/-/jobs/4464082

**Review by varunmylaraiah — APPROVED** [review]

**Comment by t-woerner** [comment]
Is it needed to duplicate all tests like this? It would be better to have the tests defined one time only.

**Comment by rjeffman** [comment]
@t-woerner, I've modified the client context tests to use original test files, which were adapted for both 'server' and 'client' context execution.

**Comment by rjeffman** [comment]
I've marked this as a "draft/WIP" as I found a few more issues with modules running with `ipaapi_context: client` (specially `ipaconfig`). Also all tests need to be modified so that at least part of them can be executed in upstream CI.

File: README-topology.md
Comment (t-woerner) [inline README-topology.md:167]: Please add the missing `no` instead.

File: plugins/modules/ipaconfig.py
Comment (t-woerner) [inline plugins/modules/ipaconfig.py:255]: Isn't this also in PR #637?

File: plugins/modules/ipapermission.py
Comment (t-woerner) [inline plugins/modules/ipapermission.py:374]: Is this needed for client context? Please add a comment.

File: plugins/module_utils/ansible_freeipa_module.py
Comment (t-woerner) [inline plugins/module_utils/ansible_freeipa_module.py:189]: How about using a client context on a client automatically if the context is not set? This way a task could be executed on a client without specifying the context.

Comment (t-woerner) [reply plugins/module_utils/ansible_freeipa_module.py:189]: Otherwise context would be a required parameter on clients always.

Comment (t-woerner) [reply plugins/module_utils/ansible_freeipa_module.py:189]: if context is None:
        if api.env.in_server:
            context = "server"
        else:
            context = "client"

File: tests/automember/test_automember_client_context.yml
Comment (t-woerner) [inline tests/automember/test_automember_client_context.yml:34]: If the context defaults to "client" on a client, this line should not be needed.

File: README-topology.md
Comment (rjeffman) [reply README-topology.md:167]: I'll add the missing "no". My text editor removed the trailing space, I didn't notice it.

File: plugins/modules/ipaconfig.py
Comment (rjeffman) [reply plugins/modules/ipaconfig.py:255]: yes, it is in PR #637. Forgot to remove it from here.

File: plugins/module_utils/ansible_freeipa_module.py
Comment (rjeffman) [reply plugins/module_utils/ansible_freeipa_module.py:189]: I was thinking on forcing explicit usage if client context is to be used. Mostly because I didn't know if all modules would run on it. I'll change it to select the proper context, and change documentation, as `ipaapi_context` will not default to `server`.

File: plugins/modules/ipapermission.py
Comment (rjeffman) [reply plugins/modules/ipapermission.py:374]: It is needed for the client context, at least, part of it. I'll add a comment.

File: plugins/module_utils/ansible_freeipa_module.py
Comment (rjeffman) [reply plugins/module_utils/ansible_freeipa_module.py:189]: fixed.

File: tests/dnsrecord/env_cleanup.yml
Comment (rjeffman) [inline tests/dnsrecord/env_cleanup.yml:1]: The patch to this file is bigger because indentation was wrong and yamllint complained about it.

File: tests/automember/test_automember_client_context.yml
Comment (rjeffman) [reply tests/automember/test_automember_client_context.yml]: fixed.

File: plugins/modules/ipapermission.py
Comment (rjeffman) [reply plugins/modules/ipapermission.py]: fixed.

File: plugins/modules/ipaconfig.py
Comment (rjeffman) [reply plugins/modules/ipaconfig.py]: fixed.

File: README-topology.md
Comment (rjeffman) [reply README-topology.md]: fixed

**Comment by rjeffman** [comment]
Due to issues with `ipaconfig` module tests, we cannot rely o the test results.

This PR is currently blocked until PR #637 is merged.

**Comment by rjeffman** [comment]
This PR should not be merged before #635.

File: README-vault.md
Comment (t-woerner) [inline README-vault.md:220]: For vault server context is not supported, right? I think we will need a different text here.

File: tests/role/env_facts.yml
Comment (t-woerner) [inline tests/role/env_facts.yml:10]: This is a fix.

File: tests/user/test_user.yml
Comment (t-woerner) [inline tests/user/test_user.yml:95]: We should change the SSH keys. `pinky@ipaserver.el81.local` might not be optimal.

File: utils/templates/test_module+member.yml.in
Comment (t-woerner) [inline utils/templates/test_module+member.yml.in:20]: Line 3 is not correct. It should be `hosts: "{{ ipa_test_host | default('ipaserver') }}"`

File: utils/templates/test_module.yml.in
Comment (t-woerner) [inline utils/templates/test_module.yml.in:10]: Line 3 is not correct. It should be hosts: "{{ ipa_test_host | default('ipaserver') }}"
Do we also need gather_facts: true ?

File: tests/hbacsvcgroup/test_hbacsvcgroup_client_context.yml
Comment (t-woerner) [inline tests/hbacsvcgroup/test_hbacsvcgroup_client_context.yml:16]: Side note: There are `name: ThisShouldNotWork`, `name: ThisShouldNotBeAdded` and `name: THIS_SHOULD_FAIL`. Are there different templates for this file?

File: tests/role/env_facts.yml
Comment (rjeffman) [reply tests/role/env_facts.yml:10]: It shouldn't be here. This should be part of another PR.

**Comment by varunmylaraiah** [comment]
@rjeffman, Tests are working perfectly without ipaapi_context: client

File: tests/user/test_user.yml
Comment (rjeffman) [reply tests/user/test_user.yml:95]: I agree. Although these are not used, other than storing the keys.

File: utils/templates/test_module+member.yml.in
Comment (rjeffman) [reply utils/templates/test_module+member.yml.in:23]: Fixed line 3 (but see below).

File: utils/templates/test_module.yml.in
Comment (rjeffman) [reply utils/templates/test_module.yml.in:13]: I prefer that we use `gather_facts: no` unless it is required. Test playbook should also serve as documentation, and showing that "this is not required for these cases" is good, in my opinion. 

Fixed line 3.

File: tests/hbacsvcgroup/test_hbacsvcgroup_client_context.yml
Comment (rjeffman) [reply tests/hbacsvcgroup/test_hbacsvcgroup_client_context.yml]: Fixed.

File: README-vault.md
Comment (rjeffman) [reply README-vault.md]: Fixed.

File: README-vault.md
Comment (t-woerner) [inline README-vault.md:220]: Please change `falure` to `failure`.

File: tests/automember/test_automember_client_context.yml
Comment (t-woerner) [inline tests/automember/test_automember_client_context.yml:17]: This made me look into the docs to see if this is supported. It is supported by the module, but there are no docs #648

File: tests/user/test_user_client_context.yml
Comment (t-woerner) [inline tests/user/test_user_client_context.yml:30]: This should be `import_playbook: test_${name}.yml`

File: tests/user/test_user_client_context.yml
Comment (t-woerner) [inline tests/user/test_user_client_context.yml:36]: This should be import_playbook: test_${name}.yml

File: tests/vault/test_vault_client_context.yml
Comment (t-woerner) [inline tests/vault/test_vault_client_context.yml:3]: Do we also need `hosts: ipaclients, ipaserver` here?

File: tests/automember/test_automember_client_context.yml
Comment (t-woerner) [inline tests/automember/test_automember_client_context.yml:3]: Do we need `ipaserver` here? It is not in all the `test_*_client_context.yml` files.

File: tests/automember/test_automember_client_context.yml
Comment (rjeffman) [reply tests/automember/test_automember_client_context.yml:17]: Thank you.

File: tests/vault/test_vault_client_context.yml
Comment (rjeffman) [reply tests/vault/test_vault_client_context.yml:3]: No, as we only need to test the failure in the case where `ipaapi_context: server`.

File: tests/automember/test_automember_client_context.yml
Comment (rjeffman) [reply tests/automember/test_automember_client_context.yml:3]: When using both, you can this file against a client and a server, using something similar to the [example inventory file for a cluster](https://github.com/freeipa/ansible-freeipa/blob/master/inventory/hosts.cluster).

File: README-vault.md
Comment (rjeffman) [reply README-vault.md]: Fixed.

File: tests/automember/test_automember_client_context.yml
Comment (rjeffman) [reply tests/automember/test_automember_client_context.yml:3]: I added both host list to all `*_client_context.yml` files.

**Review by t-woerner — APPROVED** [review]
LGTM

**Pull request merged**
